File: YeetMacro2/Resources/Raw/MacroSets/konosubaFD/scripts/farmMats.js
let done = false;
const loopPatterns = [patterns.titles.home, patterns.titles.smithy, patterns.titles.craft, patterns.skipAll.title];
const offset = macroService.CalcOffset(patterns.titles.home);

const farmMat = async (targetMats, staminaCost, numSkips) => {
	macroService.PollPattern(patterns.skipAll.material, { DoClick: true, PredicatePattern: patterns.skipAll.search });
	sleep(500);
	const filterOffResult = macroService.FindPattern(patterns.skipAll.search.filter.off);
	if (filterOffResult.IsSuccess) {
		macroService.PollPattern(patterns.skipAll.search.filter.off, { DoClick: true, PredicatePattern: patterns.skipAll.search.filter });
		sleep(500);
		const checkResult = macroService.FindPattern(patterns.skipAll.search.filter.check, { Limit: 4 });
		for (let point of checkResult.Points) {
			if (point.X < offset.X + 400.0) continue;		// skip 4 stars
			macroService.DoClick(point);
			sleep(250);
		}
		macroService.PollPattern(patterns.skipAll.search.filter.close, { DoClick: true, PredicatePattern: patterns.skipAll.search });
		sleep(500);
	}

	macroService.PollPattern(patterns.skipAll.search.select.check, { DoClick: true, InversePredicatePattern: patterns.skipAll.search.select.check });
	for (const mat of targetMats) {
		const matResult = macroService.FindPattern(mat);
		if (matResult.IsSuccess) {
			const matCheckPattern = {
				...patterns.skipAll.search.select.check,
				props: {
					...patterns.skipAll.search.select.check.props,
					path: patterns.skipAll.search.select.check.props.path + '_' + mat.props.path,
					patterns: patterns.skipAll.search.select.check.props.patterns.map(p => ({
						...p,
						rect: {
							x: matResult.Point.X - 115.0,
							y: matResult.Point.Y - 105.0,
							width: 110.0,
							height: 85.0
						},
						offsetCalcType: "None"
					})),
				}
			};
			//logger.debug(JSON.stringify(matCheckPattern, null, 2));
			macroService.PollPattern(mat, { DoClick: true, PredicatePattern: matCheckPattern, IntervalDelayMs: 1_000 });
		}
	}
	sleep(500);
	macroService.PollPattern(patterns.skipAll.search.button, { DoClick: true, PredicatePattern: patterns.skipAll.title });
	sleep(2000);

	const currentStaminaCost = macroService.GetText(patterns.skipAll.totalCost);
	if (currentStaminaCost < staminaCost) {
		macroService.PollPattern(patterns.skipAll.addStamina, { DoClick: true, PredicatePattern: patterns.stamina.prompt.recoverStamina });
		macroService.PollPattern(patterns.stamina.meat, { DoClick: true, PredicatePattern: patterns.stamina.prompt.recoverStamina2 });
		let targetStamina = macroService.GetText(patterns.stamina.target);
		while (macroService.IsRunning && targetStamina < staminaCost) {
			macroService.ClickPattern(patterns.stamina.plusOne);
			sleep(500);
			targetStamina = macroService.GetText(patterns.stamina.target);
		}
		macroService.PollPattern(patterns.stamina.prompt.recover, { DoClick: true, ClickPattern: patterns.stamina.prompt.ok, PredicatePattern: patterns.skipAll.addMaxSkips, IntervalDelayMs: 1_000 });
	}

	let maxNumSkips = macroService.GetText(patterns.skipAll.maxNumSkips);
	while (macroService.IsRunning && maxNumSkips < numSkips) {
		macroService.ClickPattern(patterns.skipAll.addMaxSkips);
		sleep(500);
		maxNumSkips = macroService.GetText(patterns.skipAll.maxNumSkips);
	}

	macroService.PollPattern(patterns.skipAll.button, { DoClick: true, PredicatePattern: patterns.skipAll.prompt.ok });
	sleep(1_000);
	macroService.PollPattern(patterns.skipAll.prompt.ok, { DoClick: true, PredicatePattern: patterns.skipAll.skipComplete });
	macroService.PollPattern(patterns.skipAll.skipComplete, { DoClick: true, ClickPattern: [patterns.skipAll.prompt.ok, patterns.branchEvent.availableNow, patterns.branchEvent.playLater, patterns.prompt.playerRankUp], PredicatePattern: patterns.skipAll.title });
};

while (macroService.IsRunning && !done) {
	const result = macroService.PollPattern(loopPatterns);
	switch (result.Path) {
		case 'titles.home':
			logger.info('farmMats: click smithy tab');
			macroService.ClickPattern(patterns.tabs.smithy);
			break;
		case 'titles.smithy':
			logger.info('farmMats: click craft');
			macroService.ClickPattern(patterns.smithy.craft);
			break;
		case 'titles.craft':
			macroService.PollPattern(patterns.smithy.craft.jewelry, { DoClick: true, PredicatePattern: patterns.smithy.craft.jewelry.list });
			sleep(500);
			macroService.PollPattern(patterns.smithy.craft.materials.archAngelFeather, { DoClick: true, PredicatePattern: patterns.smithy.prompt.howToAcquire });
			sleep(1_000);
			macroService.PollPattern(patterns.smithy.prompt.skipAll, { DoClick: true, PredicatePattern: patterns.skipAll.title });
			sleep(500);
			break;
		case 'skipAll.title':
			logger.info('farmMats: farm extreme levels');
			farmMat([patterns.skipAll.search.select.mithrilOre, patterns.skipAll.search.select.yggdrasilBranch, patterns.skipAll.search.select.platinumOre], 500, 1);
			sleep(1_000);
			logger.info('farmMats: farm skyDragonScale');
			farmMat([patterns.skipAll.search.select.skyDragonScale], 500, 3);
			done = true;
			break;
	}

	sleep(1_000);
}
logger.info('Done...');

File: YeetMacro2/Services/Scripts/KonosubaFD/farmMats.cs
using Microsoft.Extensions.Logging;
using YeetMacro2.Data.Models;
using YeetMacro2.ViewModels.NodeViewModels;

namespace YeetMacro2.Services.Scripts.KonosubaFD;
public partial class KonosubaFDScripts
{
    public string farmMats()
    {
        var done = false;
        var loopPatterns = new PatternNode[] { patterns["titles"]["home"], patterns["titles"]["smithy"], patterns["titles"]["craft"], patterns["skipAll"]["title"] };

        while (macroService.IsRunning && !done)
        {
            var result = macroService.PollPattern(loopPatterns);
            switch (result.Path)
            {
                case "titles.home":
                    logger.LogInformation("farmMats: click smithy tab");
                    macroService.ClickPattern(patterns["tabs"]["smithy"]);
                    break;
                case "titles.smithy":
                    logger.LogInformation("farmMats: click craft");
                    macroService.ClickPattern(patterns["smithy"]["craft"]);
                    break;
                case "titles.craft":
                    macroService.PollPattern(patterns["smithy"]["craft"]["jewelry"], new PollPatternFindOptions() { DoClick = true, PredicatePattern = patterns["smithy"]["craft"]["jewelry"]["list"] });
                    Thread.Sleep(500);
                    macroService.PollPattern(patterns["smithy"]["craft"]["materials"]["archAngelFeather"], new PollPatternFindOptions() { DoClick = true, PredicatePattern = patterns["smithy"]["prompt"]["howToAcquire"] });
                    Thread.Sleep(1_000);
                    macroService.PollPattern(patterns["smithy"]["prompt"]["skipAll"], new PollPatternFindOptions() { DoClick = true, PredicatePattern = patterns["skipAll"]["title"] });
                    Thread.Sleep(500);
                    break;
                case "skipAll.title":
                    logger.LogInformation("farmMats: farm extreme levels");
                    farmMat(new PatternNode[] { patterns["skipAll"]["search"]["select"]["mithrilOre"], patterns["skipAll"]["search"]["select"]["yggdrasilBranch"], patterns["skipAll"]["search"]["select"]["platinumOre"] }, 500, 1);
                    Thread.Sleep(1_000);
                    logger.LogInformation("farmMats: farm skyDragonScale");
                    farmMat(new PatternNode[] { patterns["skipAll"]["search"]["select"]["skyDragonScale"] }, 500, 3);
                    done = true;
                    break;
            }

            Thread.Sleep(1_000);
        }
        logger.LogInformation("Done...");
        return String.Empty;
    }

    public void farmMat(PatternNode[] targetMats, int staminaCost, int numSkips)
    {
        var offset = macroService.CalcOffset(patterns["titles"]["home"]);
        macroService.PollPattern(patterns["skipAll"]["material"], new PollPatternFindOptions() { DoClick = true, PredicatePattern = patterns["skipAll"]["search"] });
        Thread.Sleep(500);
        var filterOffResult = macroService.FindPattern(patterns["skipAll"]["search"]["filter"]["off"]);
        if (filterOffResult.IsSuccess)
        {
            macroService.PollPattern(patterns["skipAll"]["search"]["filter"]["off"], new PollPatternFindOptions() { DoClick = true, PredicatePattern = patterns["skipAll"]["search"]["filter"] });
            Thread.Sleep(500);
            var checkResult = macroService.FindPattern(patterns["skipAll"]["search"]["filter"]["check"], new FindOptions() { Limit = 4 });
            foreach (var point in checkResult.Points)
            {
                if (point.X < offset.X + 400.0) continue;       // skip 4 stars
                macroService.DoClick(point);
                Thread.Sleep(250);
            }
            macroService.PollPattern(patterns["skipAll"]["search"]["filter"]["close"], new PollPatternFindOptions() { DoClick = true, PredicatePattern = patterns["skipAll"]["search"] });
            Thread.Sleep(500);
        }

        macroService.PollPattern(patterns["skipAll"]["search"]["select"]["check"], new PollPatternFindOptions() { DoClick = true, InversePredicatePattern = patterns["skipAll"]["search"]["select"]["check"] });
        foreach (var mat in targetMats)
        {
            var matResult = macroService.FindPattern(mat);
            if (matResult.IsSuccess)
            {
                var matCheckPattern = PatternNodeManagerViewModel.CloneNode(patterns["skipAll"]["search"]["select"]["check"]);
                matCheckPattern.Path += $"_{mat.Path}";
                foreach (var p in matCheckPattern.Patterns)
                {
                    p.Rect = new Rect(matResult.Point.X - 115.0, matResult.Point.Y - 105.0, 110.0, 85.0);
                    p.OffsetCalcType = OffsetCalcType.None;
                }
                macroService.PollPattern(mat, new PollPatternFindOptions() { DoClick = true, PredicatePattern = matCheckPattern, IntervalDelayMs = 1_000 });
            }
        }

        Thread.Sleep(500);
        macroService.PollPattern(patterns["skipAll"]["search"]["button"], new PollPatternFindOptions() { DoClick = true, PredicatePattern = patterns["skipAll"]["title"] });
        Thread.Sleep(2000);

        var currentStaminaCost = int.Parse(macroService.GetText(patterns["skipAll"]["totalCost"]));
        if (currentStaminaCost < staminaCost)
        {
            macroService.PollPattern(patterns["skipAll"]["addStamina"], new PollPatternFindOptions() { DoClick = true, PredicatePattern = patterns["stamina"]["prompt"]["recoverStamina"] });
            macroService.PollPattern(patterns["stamina"]["meat"], new PollPatternFindOptions() { DoClick = true, PredicatePattern = patterns["stamina"]["prompt"]["recoverStamina2"] });
            var targetStamina = int.Parse(macroService.GetText(patterns["stamina"]["target"]));
            while (macroService.IsRunning && targetStamina < staminaCost)
            {
                macroService.ClickPattern(patterns["stamina"]["plusOne"]);
                Thread.Sleep(500);
                targetStamina = int.Parse(macroService.GetText(patterns["stamina"]["target"]));
            }
            macroService.PollPattern(patterns["stamina"]["prompt"]["recover"], new PollPatternFindOptions() { DoClick = true, ClickPattern = patterns["stamina"]["prompt"]["ok"], PredicatePattern = patterns["skipAll"]["addMaxSkips"], IntervalDelayMs = 1_000 });
        }

        var maxNumSkips = int.Parse(macroService.GetText(patterns["skipAll"]["maxNumSkips"]));
        while (macroService.IsRunning && maxNumSkips < numSkips)
        {
            macroService.ClickPattern(patterns["skipAll"]["addMaxSkips"]);
            Thread.Sleep(500);
            maxNumSkips = int.Parse(macroService.GetText(patterns["skipAll"]["maxNumSkips"]));
        }

        macroService.PollPattern(patterns["skipAll"]["button"], new PollPatternFindOptions() { DoClick = true, PredicatePattern = patterns["skipAll"]["prompt"]["ok"] });
        Thread.Sleep(1_000);
        macroService.PollPattern(patterns["skipAll"]["prompt"]["ok"], new PollPatternFindOptions() { DoClick = true, PredicatePattern = patterns["skipAll"]["skipComplete"] });
        macroService.PollPattern(patterns["skipAll"]["skipComplete"], new PollPatternFindOptions() { DoClick = true, ClickPattern = new PatternNode[] { patterns["skipAll"]["prompt"]["ok"], patterns["branchEvent"]["availableNow"], patterns["branchEvent"]["playLater"], patterns["prompt"]["playerRankUp"] }, PredicatePattern = patterns["skipAll"]["title"] });
    }
}